File: 347-top-k-frequent-elements/top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int,int> mp;
        for(int it:nums) mp[it]++;
        vector<vector<int>> freq(nums.size()+1);
        for(auto& i : mp){
            freq[i.second].push_back(i.first);
        }
        vector<int> ans;
        for(int i = freq.size()-1;i>-1;i--){
            for(int j : freq[i]){
                ans.push_back(j);
                k--;
                if(k<1) return ans;
            }
        }
        return ans;
    }
};

File: 347-top-k-frequent-elements/top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> mp = new HashMap<>();
        for(int i : nums)
        {
            mp.put(i,mp.getOrDefault(i,0)+1);
        }
        List<Integer>[] freq = new ArrayList[nums.length+1];
        for(Map.Entry<Integer,Integer> entry : mp.entrySet())
        {
            int kk = entry.getKey();
            int v = entry.getValue();
            if(freq[v]==null) freq[v]=new ArrayList<>();
            freq[v].add(kk);
        }
        int[] ans = new int[k];
        int i =0;
        int j = freq.length-1;
        while(j>0)
        {
            if(freq[j]!=null)
            {
                for(int t : freq[j])
                {
                    ans[i]=t;
                    i++;
                    k--;
                    if(k==0) return ans;
                    
                }
            }
            j--;
        }
        return ans;
    }
}